**Using this template**

1. Change notebook title to assignment title you working on, eg. `Graded Challenge 1 or Milestone 1`
2. Put your identity such as name and batch number below notebook title.
3. Write your description before do any work. State your purpose of this assignment and what you are trying to achieve. 
4. Look at the rubric of this assignment and make sure you understand the criteria.
5. Code your solution in the cell provided below the working area section.
6. State your conlusions, findings, and any other relevant information in the cell provided below the conclusions sections.
7. Save this notebook and rename it to assignment title you working on  eg. `Graded Challenge 1 or Milestone 1`
8. Push your assignment before deadline.
9. Good Luck!

# Notebook Title

Graded Challenge 2

Nama: Gusti Ayu Dewi Puspa Kartikasari

Batch: 012

## Description

### Objective
Do the steps stated in the GC problem and answer the questions


### Problem Statement
Problem 1 - EIGENDECOMPOSITION
a. Mengapa dalam kasus ini kita menerapkan konsep Eigendecomposition?

b. Dari pengamatanmu, mengapa pada akhirnya kita menggunakan dataframe df_matrix_square dibandingkan df_matrix untuk dilakukan proses Eigendecomposition? Berikan alasan yang logis sesuai dengan konsep Eigendecomposition.

c. Jika dilihat dari PageRanknya, dimana nilai PageRank paling besar merupakan stasiun yang paling favorit. Apa insight yang akan kamu berikan?

-------------------------------------------

Problem 2 - INTEGRAL

Hitung luas area yang dilingkup stasiun awal menggunakan library Scipy. Hasilnya bersatuan square degree.

a. Terdapat 2 jenis integral yaitu definite dan indefinite. Jenis integral apa yang kamu gunakan untuk kasus ini? Berikan alasanmu!

b. Terdapat dua metode dalam perhitungan integral yaitu simbolik dan numerik. Mana yang kamu gunakan untuk kasus ini dan mengapa?

c. Apakah diperlukan mendefinisikan fungsi matematis ( seperti f(x) = x^2+2 ) pada kasus ini? Berikan penjelasan dan alasanmu!

## Working Area

```txt
Put your code, analysis, everything below this line

Make sure to check everything the rubric requires before you submit
```

### SQL CODE

select trips.trip_id,trips.start_date,trips.start_station_name,trips.end_station_name,trips.start_station_latitude,trips.start_station_longitude,trips.end_station_latitude,trips.end_station_longitude,info.region_id
from `bigquery-public-data.san_francisco_bikeshare.bikeshare_station_info` as info
join `bigquery-public-data.san_francisco_bikeshare.bikeshare_trips` as trips on info.station_id = trips.start_station_id
WHERE trips.start_station_name != trips.end_station_name
AND trips.start_station_latitude IS NOT NULL
AND trips.start_station_longitude IS NOT NULL
AND trips.end_station_latitude IS NOT NULL
AND trips.end_station_longitude IS NOT NULL
AND trips.start_date BETWEEN '2017-04-24' AND '2017-09-24';

### Import library

In [1]:
import pandas as pd
import numpy as np

## Problem 1

### Loading data

In [2]:
df = pd.read_csv(r'h8dsft_P0GC2_Set_1_Gusti_Ayu_Dewi_Puspa_Kartikasari.csv')
df.head(3)

trip_id               start_date  start_station_name  \
0  22232017090510434500  2017-09-05 10:43:45 UTC  12th St at 4th Ave   
1  10412017072215351700  2017-07-22 15:35:17 UTC  12th St at 4th Ave   
2  12232017080409490700  2017-08-04 09:49:07 UTC  12th St at 4th Ave   

  end_station_name  start_station_latitude  start_station_longitude  \
0        Snow Park               37.795812              -122.255555   
1        Snow Park               37.795812              -122.255555   
2        Snow Park               37.795812              -122.255555   

   end_station_latitude  end_station_longitude  region_id  
0             37.807813            -122.264496         12  
1             37.807813            -122.264496         12  
2             37.807813            -122.264496         12

### Data Processing

1. Buatlah dua variable baru yang bernama start_stations dan end_stations yang masing-masing berisikan list nama stasiun awal dan stasiun akhir. Pastikan tidak ada nama stasiun yang duplikat di masing-masing variable

In [3]:
start_stations = np.unique(df.start_station_name)
end_stations = np.unique(df.end_station_name)

2. Deteksi stasiun-stasiun yang hanya terdapat di salah satu variable dan tidak di keduanya (exclusive-or/xor) menggunakan np.setxor1d(array1,array2).

In [4]:
exclusive_stations = np.setxor1d(start_stations,end_stations)
exclusive_stations

array(['2nd St at S Park St', '5th St at Brannan St',
       '5th St at Howard St', '7th St at Brannan St',
       'Franklin St at 14th St', 'Golden Gate Ave at Hyde St',
       'Irwin St at 8th St', 'Jackson St at 14th St',
       'Laguna St at McAllister St', 'Palm St at Willow St',
       'San Francisco Caltrain Station 2  (Townsend St at 4th St)',
       'Webster St at 17th St'], dtype=object)

3. Buat dataframe yang berisikan cross-tabulation antara start_stations dan end_stations menggunakan pd.crosstab(pd.Series1,pd.Series2). Hasil dari langkah ini akan menghasilkan dataframe yang menginfokan berapa jumlah/frekuensi perjalanan dari start station tertentu ke end station tertentu.

In [5]:
df_matrix = pd.crosstab(df.start_station_name,df.end_station_name)
df_matrix.head(5)

end_station_name       10th Ave at E 15th St  10th St at Fallon St  \
start_station_name                                                   
10th Ave at E 15th St                      0                     6   
10th St at Fallon St                      10                     0   
11th St at Bryant St                       0                     0   
11th St at Natoma St                       0                     0   
12th St at 4th Ave                         0                   152   

end_station_name       11th St at Bryant St  11th St at Natoma St  \
start_station_name                                                  
10th Ave at E 15th St                     0                     0   
10th St at Fallon St                      0                     0   
11th St at Bryant St                      0                     9   
11th St at Natoma St                      9                     0   
12th St at 4th Ave                        0                     0   

end_station_name       12th St at 4th Ave  12th St at 6th Ave  \
start_station_name                                              
10th Ave at E 15th St                   0                   1   
10th St at Fallon St                  119                  19   
11th St at Bryant St                    0                   0   
11th St at Natoma St                    0                   0   
12th St at 4th Ave                      0                   4   

end_station_name       14th St at Filbert St  14th St at Mandela Pkwy  \
start_station_name                                                      
10th Ave at E 15th St                      0                        0   
10th St at Fallon St                       0                        4   
11th St at Bryant St                       0                        0   
11th St at Natoma St                       0                        0   
12th St at 4th Ave                         0                        0   

end_station_name       14th St at Mission St  16th St Mission BART  \
start_station_name                                                   
10th Ave at E 15th St                      0                     0   
10th St at Fallon St                       0                     0   
11th St at Bryant St                       4                     1   
11th St at Natoma St                       1                     0   
12th St at 4th Ave                         0                     0   

end_station_name                               ...                            \
start_station_name                             ...                             
10th Ave at E 15th St                          ...                             
10th St at Fallon St                           ...                             
11th St at Bryant St                           ...                             
11th St at Natoma St                           ...                             
12th St at 4th Ave                             ...                             

end_station_name       Victoria Manalo Draves Park  Washington St at 8th St  \
start_station_name                                                            
10th Ave at E 15th St                            0                        0   
10th St at Fallon St                             0                        7   
11th St at Bryant St                             5                        0   
11th St at Natoma St                             6                        0   
12th St at 4th Ave                               0                        1   

end_station_name       Washington St at Kearny St  Webster St at 17th St  \
start_station_name                                                         
10th Ave at E 15th St                           0                      0   
10th St at Fallon St                            0                      8   
11th St at Bryant St                            0                      0   
11th St at Natoma St                            0                      

In [6]:
df_matrix.shape

(218, 230)

Note: df_matrix tidak memiliki jumlah kolom dan baris yang sama

4. Buatlah salinan dari data yang di-load ke dalam variable bernama df_copy.

In [7]:
df_copy = df.copy()


5. Filter df_copy dimana tidak ada start_station_name dan end_station_name yang termasuk dalam list exclusive_stations.

In [8]:
df_copy = df_copy[
        (~df_copy['start_station_name'].isin(exclusive_stations)) &
        (~df_copy['end_station_name'].isin(exclusive_stations))
]
df_copy.head(4)

trip_id               start_date  start_station_name  \
0  22232017090510434500  2017-09-05 10:43:45 UTC  12th St at 4th Ave   
1  10412017072215351700  2017-07-22 15:35:17 UTC  12th St at 4th Ave   
2  12232017080409490700  2017-08-04 09:49:07 UTC  12th St at 4th Ave   
3   4992017082809165300  2017-08-28 09:16:53 UTC  12th St at 4th Ave   

  end_station_name  start_station_latitude  start_station_longitude  \
0        Snow Park               37.795812              -122.255555   
1        Snow Park               37.795812              -122.255555   
2        Snow Park               37.795812              -122.255555   
3        Snow Park               37.795812              -122.255555   

   end_station_latitude  end_station_longitude  region_id  
0             37.807813            -122.264496         12  
1             37.807813            -122.264496         12  
2             37.807813            -122.264496         12  
3             37.807813            -122.264496         12

6. Lakukan langkah 3 untuk variable df dan simpan ke dalam variable bernama df_matrix_square. Pastikan jumlah rows dan columns sama, kalau tidak ulangi dari langkah 4.

In [9]:
df_matrix_square = pd.crosstab(df_copy.start_station_name,df_copy.end_station_name)
df_matrix_square.head()

end_station_name       10th Ave at E 15th St  10th St at Fallon St  \
start_station_name                                                   
10th Ave at E 15th St                      0                     6   
10th St at Fallon St                      10                     0   
11th St at Bryant St                       0                     0   
11th St at Natoma St                       0                     0   
12th St at 4th Ave                         0                   152   

end_station_name       11th St at Bryant St  11th St at Natoma St  \
start_station_name                                                  
10th Ave at E 15th St                     0                     0   
10th St at Fallon St                      0                     0   
11th St at Bryant St                      0                     9   
11th St at Natoma St                      9                     0   
12th St at 4th Ave                        0                     0   

end_station_name       12th St at 4th Ave  12th St at 6th Ave  \
start_station_name                                              
10th Ave at E 15th St                   0                   1   
10th St at Fallon St                  119                  19   
11th St at Bryant St                    0                   0   
11th St at Natoma St                    0                   0   
12th St at 4th Ave                      0                   4   

end_station_name       14th St at Filbert St  14th St at Mandela Pkwy  \
start_station_name                                                      
10th Ave at E 15th St                      0                        0   
10th St at Fallon St                       0                        4   
11th St at Bryant St                       0                        0   
11th St at Natoma St                       0                        0   
12th St at 4th Ave                         0                        0   

end_station_name       14th St at Mission St  16th St Mission BART  \
start_station_name                                                   
10th Ave at E 15th St                      0                     0   
10th St at Fallon St                       0                     0   
11th St at Bryant St                       4                     1   
11th St at Natoma St                       1                     0   
12th St at 4th Ave                         0                     0   

end_station_name                               ...                            \
start_station_name                             ...                             
10th Ave at E 15th St                          ...                             
10th St at Fallon St                           ...                             
11th St at Bryant St                           ...                             
11th St at Natoma St                           ...                             
12th St at 4th Ave                             ...                             

end_station_name       Valencia St at Cesar Chavez St  \
start_station_name                                      
10th Ave at E 15th St                               0   
10th St at Fallon St                                0   
11th St at Bryant St                                2   
11th St at Natoma St                                1   
12th St at 4th Ave                                  0   

end_station_name       Victoria Manalo Draves Park  Washington St at 8th St  \
start_station_name                                                            
10th Ave at E 15th St                            0                        0   
10th St at Fallon St                             0                        7   
11th St at Bryant St                             5                        0   
11th St at Natoma St                             6                        0   
12th St at 4th Ave                               0                        1   

end_station_name       Washington St at Kearny St  

In [10]:
df_matrix_square.shape

(218, 218)

Note: df_matrix_square memiliki jumlah kolom dan baris yang sama

7. Jumlahkan nilai kolom masing-masing rows pada df_matrix_square menggunakan method sum(axis=1) dan simpan ke dalam variable bernama total_trips.

In [11]:
total_trips=df_matrix_square.sum(axis=1)
total_trips.head()

start_station_name
10th Ave at E 15th St     62
10th St at Fallon St     630
11th St at Bryant St     400
11th St at Natoma St     409
12th St at 4th Ave       386
dtype: int64

In [42]:
total_trips.max()

6886

In [38]:
total_trips[149]

332

8. Bagi setiap kolom df_matrix_square dengan menggunakan method .div(), dengan inputan parameter total_trips dan axis='rows'. Masukkan hasilnya ke variable weighted_matrix_df.

In [28]:
weighted_matrix_df = df_matrix_square.div(total_trips, axis='rows')
weighted_matrix_df.head(10)

end_station_name         10th Ave at E 15th St  10th St at Fallon St  \
start_station_name                                                     
10th Ave at E 15th St                 0.000000              0.096774   
10th St at Fallon St                  0.015873              0.000000   
11th St at Bryant St                  0.000000              0.000000   
11th St at Natoma St                  0.000000              0.000000   
12th St at 4th Ave                    0.000000              0.393782   
12th St at 6th Ave                    0.000000              0.397590   
14th St at Filbert St                 0.000000              0.000000   
14th St at Mandela Pkwy               0.000000              0.046784   
14th St at Mission St                 0.000000              0.000000   
16th St Mission BART                  0.000000              0.000000   

end_station_name         11th St at Bryant St  11th St at Natoma St  \
start_station_name                                                    
10th Ave at E 15th St                0.000000              0.000000   
10th St at Fallon St                 0.000000              0.000000   
11th St at Bryant St                 0.000000              0.022500   
11th St at Natoma St                 0.022005              0.000000   
12th St at 4th Ave                   0.000000              0.000000   
12th St at 6th Ave                   0.000000              0.000000   
14th St at Filbert St                0.000000              0.000000   
14th St at Mandela Pkwy              0.000000              0.000000   
14th St at Mission St                0.006983              0.001397   
16th St Mission BART                 0.000000              0.000000   

end_station_name         12th St at 4th Ave  12th St at 6th Ave  \
start_station_name                                                
10th Ave at E 15th St              0.000000            0.016129   
10th St at Fallon St               0.188889            0.030159   
11th St at Bryant St               0.000000            0.000000   
11th St at Natoma St               0.000000            0.000000   
12th St at 4th Ave                 0.000000            0.010363   
12th St at 6th Ave                 0.012048            0.000000   
14th St at Filbert St              0.000000            0.000000   
14th St at Mandela Pkwy            0.000000            0.000000   
14th St at Mission St              0.000000            0.000000   
16th St Mission BART               0.000000            0.000000   

end_station_name         14th St at Filbert St  14th St at Mandela Pkwy  \
start_station_name                                                        
10th Ave at E 15th St                      0.0                 0.000000   
10th St at Fallon St                       0.0                 0.006349   
11th St at Bryant St                       0.0                 0.000000   
11th St at Natoma St                       0.0                 0.000000   
12th St at 4th Ave                         0.0                 0.000000   
12th St at 6th Ave                         0.0                 0.000000   
14th St at Filbert St                      0.0                 0.000000   
14th St at Mandela Pkwy                    0.0                 0.000000   
14th St at Mission St                      0.0                 0.000000   
16th St Mission BART                       0.0                 0.000000   

end_station_name         14th St at Mission St  16th St Mission BART  \
start_station_name                                                     
10th Ave at E 15th St                 0.000000                0.0000   
10th St at Fallon St                  0.000000                0.0000   
11th St at Bryant St                  0.010000                0.0025   
11th St at Natoma St                  0.002445                0.0000   
12th St at 4th Ave                    0.000000                0.0000   
12th St at 6th Ave                    0.000000                0.0000   


In [40]:
start_stations[149]

'Paseo De San Antonio at 2nd St'

In [41]:
df_matrix_square.index[149]

'Paseo De San Antonio at 2nd St'

In [30]:
total_trips2 = weighted_matrix_df.sum(axis=0)
total_trips2

end_station_name
10th Ave at E 15th St                                    0.242852
10th St at Fallon St                                     1.911026
11th St at Bryant St                                     0.611903
11th St at Natoma St                                     0.570694
12th St at 4th Ave                                       0.560301
12th St at 6th Ave                                       0.254922
14th St at Filbert St                                    0.005547
14th St at Mandela Pkwy                                  0.759796
14th St at Mission St                                    0.774975
16th St Mission BART                                     0.007121
16th St at Prosper St                                    0.339369
17th & Folsom Street Park (17th St at Folsom St)         0.510462
17th St at Dolores St                                    0.633473
17th St at Valencia St                                   1.604939
18th St at Noe St                                        1.

9. Konversikan weighted_matrix_df ke numpy array

In [13]:
weighted_matrix_df = np.array(weighted_matrix_df)

10. Hitung eigen value dan eigen vectormenggunakan library numpy.

In [14]:
eig_val,eig_vec = np.linalg.eig(np.array(weighted_matrix_df))


11. Score PageRank terdapat pada eigen vector yang eigen valuenya paling tinggi. Simpan eigen vector tersebut ke dalam variable bernama PR_Scores.

In [15]:
print('posisi eigen value terbesar adalah ', np.argmax(eig_val), ' : ' , np.abs(eig_val[np.argmax(eig_val)]))

posisi eigen value terbesar adalah  3  :  0.9999999999999998


In [31]:
PR_Scores=eig_vec[np.argmax(eig_val)]

12. Buat dataframe baru bernama PageRank_df dengan kolom 'start_station' yang berisikan value variable start_stations dan kolom score yang berisikan value variable PR_Scores

In [43]:
PageRank_df = pd.DataFrame({df_matrix_square.index,'score': PR_Scores})

SyntaxError: invalid syntax (<ipython-input-43-56d1a4fceb4a>, line 1)

13. Tampilkan head dari dataframe PageRank_df!

In [33]:
PageRank_df.head(10)

start_station                         score
0    10th Ave at E 15th St     (-0.00953439982546328+0j)
1     10th St at Fallon St     (-0.07541666366287934+0j)
2     11th St at Bryant St  (-4.4192080039343425e-05+0j)
3     11th St at Natoma St      (0.06772854615313222+0j)
4       12th St at 4th Ave      (0.09521660703553411+0j)
5       12th St at 6th Ave   (-6.715126885834749e-05+0j)
6    14th St at Filbert St  (-4.7863344125090085e-06+0j)
7  14th St at Mandela Pkwy    (5.108358220618956e-07+0j)
8    14th St at Mission St  (-1.2051035170068618e-05+0j)
9     16th St Mission BART   (-6.437477046422507e-05+0j)

In [34]:
print('\nposisi eigen vector terbesar: ', np.argmax(eig_vec[3,:]), ', stasiun ',start_stations[np.argmax(eig_vec[3,:])])



posisi eigen vector terbesar:  149 , stasiun  Paseo De San Antonio at 2nd St


In [35]:
PageRank_df.sort_values(by="score", ascending=False).head(10)

start_station  \
149  Paseo De San Antonio at 2nd St   
150            Post St at Kearny St   
181         Shattuck Ave at 51st St   
180     Scott St at Golden Gate Ave   
135        Market St at Franklin St   
136            Market St at Park St   
4                12th St at 4th Ave   
145       Morrison Ave at Julian St   
146                   Mosswood Park   
122           Laguna St at Hayes St   

                                            score  
149  (0.14840206406213458+0.0015457255935450165j)  
150  (0.14840206406213458-0.0015457255935450165j)  
181     (0.1102001151770241+0.06214915102293913j)  
180     (0.1102001151770241-0.06214915102293913j)  
135    (0.09872497866358776+0.06804483351673378j)  
136    (0.09872497866358776-0.06804483351673378j)  
4                        (0.09521660703553411+0j)  
145    (0.09287949947420257+0.03824050281558655j)  
146    (0.09287949947420257-0.03824050281558655j)  
122                       (0.0801377675017742+0j)

In [22]:
[[np.argmax(PR_Scores)]]

start_station  \
149  Paseo De San Antonio at 2nd St   

                                            score  
149  (0.14840206406213458+0.0015457255935450165j)

In [36]:
[PR_Scores[111]]

[(0.01300561579826139-0.025258678688105666j)]

In [26]:
print(np.abs(PageRank_df.score.iloc[149]))
print(np.abs(PageRank_df.score.iloc[111]))

0.14841011382487526
0.028410330715444857


Note: Nilai page rank terbesar di raih oleh Paseo De San Antonio at 2nd St, stasiun ini merupakain stasiun favorit warga San Fransisco

### Pertanyaan Problem 1
a. Mengapa dalam kasus ini kita menerapkan konsep Eigendecomposition?

Eigendecomposition digunakan untuk mencari page rank eigen, dimana stasiun yang memiliki nilai page rank terbesar dapat diidentifikasi sebagai stasiun yang paling sering dikunjungi atau terhubung dengan stasiun lain (favorit warga san fransisco).

b. Dari pengamatanmu, mengapa pada akhirnya kita menggunakan dataframe df_matrix_square dibandingkan df_matrix untuk dilakukan proses Eigendecomposition? Berikan alasan yang logis sesuai dengan konsep Eigendecomposition.

df_matrix memiliki ukuran 218, 230 yang berarti baris dan kolom tidak memiliki jumlah yang sama dikarenakan masih adanya exclusive stattion. Eigen decomposition merupakan proses dekomposisis matrix square (jumlah kolom dan baris sama) menjadi beberapa pasang eigen values dan eigen vector, sehingga df_matrix tidak dapat digunakan.

c. Jika dilihat dari PageRanknya, dimana nilai PageRank paling besar merupakan stasiun yang paling favorit. Apa insight yang akan kamu berikan?

Nilai PageRank yang paling besar dimiliki oleh stasiun  Paseo De San Antonio at 2nd Stn hal ini menunjukkan bahwa stasiun tersebut memiliki jumlah trip yang besar di bandingkan stasiun lain. Stasiun tersebut dapat menjadi fokus pembangunan dari kapasitas hingga fasilitas kedepannya.

# PROBLEM2

1. Buat variable baru bernama df_coor_stations yang berisikan data yang sudah di-load dimana kolom region_id = 3. Pastikan tidak ada data yang duplikat dan hanya menyertakan kolom start_station_name,start_station_latitude, dan start_station_longitude. Urutkan pula dataframe tersebut berdasarkan start_station_longitude dari nilai terkecil hingga terbesar.

In [ ]:
df_coor_stations = df.query("region_id ==3")[['start_station_name', 'start_station_latitude', 'start_station_longitude']].sort_values(by="start_station_longitude", ascending=True).drop_duplicates()
df_coor_stations.head()

start_station_name  start_station_latitude  \
110518         Central Ave at Fell St               37.773311   
26279       McAllister St at Baker St               37.777416   
163669  O'Farrell St at Divisadero St               37.782405   
91408          Broderick St at Oak St               37.773063   
110530         Grove St at Divisadero               37.775946   

        start_station_longitude  
110518              -122.444293  
26279               -122.441838  
163669              -122.439446  
91408               -122.439078  
110530              -122.437777

2. Plot grafik area stasiun-stasiun di San Fransisco menggunakan method .plot(). Catatan: Gunakan arahan berikut dalam penggunaan method .plot(), .plot(x='nama kolom untuk sumbu x',y='nama kolom untuk sumbu y',figsize=(20,8)). Logitude merupakan sumbu x dan Latitude merupakan sumbu y.

In [ ]:
import matplotlib.pyplot as plt
plt.figure(figsize=(20,8))
plt.plot(df_coor_stations.start_station_longitude,df_coor_stations.start_station_latitude)

3. Hitung luas area yang dilingkup stasiun awal menggunakan library Scipy. Hasilnya bersatuan square degree.

In [ ]:
from scipy.integrate import trapz
area = trapz(df_coor_stations.start_station_latitude,df_coor_stations.start_station_longitude)
print('Luas area:', area)

Luas area: 2.1226954522939407


Luas area San Frasisco = 2.12 derajat kuadrat

### Pertanyaan Problem 2

a. Terdapat 2 jenis integral yaitu definite dan indefinite. Jenis integral apa yang kamu gunakan untuk kasus ini? Berikan alasanmu!
Kasus ini menggunakan definite integral, dikarenakan memiliki data yang jelas (latitude sebagai y dan longitude sebagai x) sebagai batas tertentu pada perhitungannya.

b. Terdapat dua metode dalam perhitungan integral yaitu simbolik dan numerik. Mana yang kamu gunakan untuk kasus ini dan mengapa?

Perhitungan ini menggunakan metode numerik, karena hasil yang ingin didapat adalah dalam bentuk angka bukan fungsi. Metode numeric memiliki output dalam satuan angka seperti pada kasus ini yakni luas area sebesar 2.12 derajat kuadrat, sedangkan symbolic akan menghasilkan hasil tanpa batasan tertentu yakni fungsi (f(x)).

c. Apakah diperlukan mendefinisikan fungsi matematis ( seperti f(x) = x^2+2 ) pada kasus ini? Berikan penjelasan dan alasanmu!

Tidak perlu, dikarenakan data yang digunakan memiliki titik/ koordinat yang jelas sehingga dapat menggunakan definite integral - numeric - trapz. Dimana trapz ini menggunakan input dalam bentuk array (titik/koordinat) yang selanjutnya digunakan untuk menghitung luas grafik dalam bentuk penggabungan banyak trapesium mengikuti grafik. 

## Conclusions, Assumptions, Overall Analysis

1. Nilai PageRank yang paling besar dimiliki oleh stasiun  Paseo De San Antonio at 2nd Stn hal ini menunjukkan bahwa stasiun tersebut memiliki jumlah trip yang besar di bandingkan stasiun lain (stasiun favorit).

2. Luas area San Frasisco = 2.12 derajat kuadrat